rio/simulator/control.ml

```diff
@@ -5,6 +5,16 @@ type t = {
   z_out : State.t -> Packet.t -> State.t;
 }
 
+let find_mapi f lst =
+  let rec aux i = function
+    | [] -> None
+    | x :: xs -> (
+        match f i x with
+        | Some _ as result -> result
+        | None -> aux (i + 1) xs)
+  in
+  aux 0 lst
+
```

Comment:
I'll let you and @KabirSamsi chat to see if this copy-paste can be massaged away.